crates/bevy_ecs/src/query/iter.rs

```diff
@@ -2540,21 +2570,21 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIterationCursor<'w, 's, D, F> {
                     unsafe { self.table_entities.get_unchecked(self.current_row as usize) };
                 // SAFETY: The row is less than the u32 len, so it must not be max.
                 let row = unsafe { TableRow::new(NonMaxU32::new_unchecked(self.current_row)) };
-                if !F::filter_fetch(&query_state.filter_state, &mut self.filter, *entity, row) {
-                    self.current_row += 1;
-                    continue;
-                }
-
-                // SAFETY:
-                // - set_table was called prior.
-                // - `current_row` must be a table row in range of the current table,
-                //   because if it was not, then the above would have been executed.
-                // - fetch is only called once for each `entity`.
-                let item =
-                    unsafe { D::fetch(&query_state.fetch_state, &mut self.fetch, *entity, row) };
-
                 self.current_row += 1;
-                return Some(item);
+
+                if F::filter_fetch(&query_state.filter_state, &mut self.filter, *entity, row)
```

Comment:
nit: I prefer the previous way this was written with the continue. I think with this pr you'd need multiple? having the safety comment being in this `if` is awkward to read.